tests/test_otel.py

```diff
@@ -0,0 +1,524 @@
+
+    with socket.create_connection(("127.0.0.1", port)) as sock:
+        if scheme == "https":
+            ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+            ctx.check_hostname = False
+            ctx.verify_mode = ssl.CERT_NONE
+            with ctx.wrap_socket(sock, server_hostname="127.0.0.1") as ssock:
+                recv = do_get(ssock, path)
+        else:
+            recv = do_get(sock, path)
+    return recv
+
+
+@pytest.fixture
+def batches(trace_service, http_ver):
+    return trace_service.spans[http_ver * 3 : http_ver * 3 + 3]
+
+
+@pytest.fixture
+def span(batches, idx):
+    return batches[idx // 10][0].scope_spans[0].spans[idx % 10]
+
+
+@pytest.fixture
+def headers(http_ver, idx):
+    if http_ver:
+        return _headers.get(f"http{http_ver}")[idx]
+
+
+@pytest.fixture(scope="module")
+def _otelcol(pytestconfig, testdir, logger):
+    if pytestconfig.option.otelcol is None:
+        yield
+        return
+
+    (testdir / "otel-config.yaml").write_text(
+        """receivers:
+  otlp:
+    protocols:
+      grpc:
+        endpoint: 127.0.0.1:14317
+
+exporters:
+  otlp/auth:
+    endpoint: 127.0.0.1:24317
+    tls:
+      insecure: true
+
+service:
+  pipelines:
+    traces:
+      receivers:
+        - otlp
+      exporters:
+        - otlp/auth"""
+    )
+    logger.info("Starting otelcol at 127.0.0.1:14317...")
+    proc = subprocess.Popen(
+        [pytestconfig.option.otelcol, "--config", testdir / "otel-config.yaml"]
+    )
+    assert proc.poll() is None, "Can't start otelcol"
+    time.sleep(1)  # give some time to get ready
+    yield
+    logger.info("Stopping otelcol...")
+    proc.terminate()
+    try:
+        proc.wait(timeout=15)
+    except subprocess.TimeoutExpired:
+        proc.kill()
+
+
+@pytest.fixture
+def response(logger, http_ver, scheme, path, headers):
+    port = 18443 if scheme == "https" else 18080
+    if http_ver == 0:
+        return simple_client(scheme, port, path, logger)
+    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    with niquests.Session(multiplexed=True) as s:
+        if http_ver == 3:
+            assert scheme == "https", "Only https:// URLs are supported."
+            s.quic_cache_layer.add_domain("127.0.0.1", port)
+        resp = s.get(
+            f"{scheme}://127.0.0.1:{port}{path}", headers=headers, verify=False
+        )
+    collect_headers(resp.headers, http_ver)
+    return resp.text
+
+
+@pytest.mark.usefixtures("trace_service", "_otelcol", "nginx")
+@pytest.mark.parametrize(
+    ("nginx_config", "http_ver", "scheme"),
+    [
+        ({"name": "test_http0", "mode": ""}, 0, "http"),
+        ({"name": "test_http1", "mode": ""}, 1, "http"),
+        ({"name": "test_http2", "mode": "ssl http2"}, 2, "https"),
+        ({"name": "test_http3", "mode": "quic"}, 3, "https"),
+    ],
+    indirect=["nginx_config"],
+    ids=["http 0.9", "http 1.1", "http 2.0 ssl", "http 3.0 quic"],
+    scope="module",
+)
+class TestOTelGenerateSpans:
```

Comment:
As discussed "offline", this test is pretty useless and should be removed.

Reply:
Fixed.